README
NAME
    URI::Amazon::APA - URI to access Amazon Product Advertising API

VERSION
    $Id: README,v 0.1 2009/05/10 11:04:33 dankogai Exp dankogai $

SYNOPSIS
      # self-explanatory
      use strict;
      use warnings;
      use URI::Amazon::APA;
      use LWP::UserAgent;
      use XML::Simple;
      use YAML::Syck;

      use URI::Amazon::APA; # instead of URI
      my $u = URI::Amazon::APA->new('http://webservices.amazon.com/onca/xml');
      $u->query_form(
        Service     => 'AWSECommerceService',
        Operation   => 'ItemSearch',
        Title       => shift || 'Perl',
        SearchIndex => 'Books',
      );
      $u->sign(
        key    => $public_key,
        secret => $private_key,
      );

      my $ua = LWP::UserAgent->new;
      my $r  = $ua->get($u);
      if ( $r->is_success ) {
        print YAML::Syck::Dump( XMLin( $r->content ) );
      }
      else {
        print $r->status_line, $r->as_string;
      }

EXPORT
    None.

METHODS
    This adds the following methods to URI object

  sign
    Sings the URI accordingly to the Amazon Product Advertising API.

      $u->sign(
        key    => $public_key,
        secret => $private_key,
      );

  signature
    Checks the signature within the URI;

      print "The signature is " : $u->signature;

AUTHOR
    Dan Kogai, "<dankogai at dan.co.jp>"

BUGS
    Please report any bugs or feature requests to "bug-uri-amazon-apa at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=URI-Amazon-APA>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc URI::Amazon::APA

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=URI-Amazon-APA>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/URI-Amazon-APA>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/URI-Amazon-APA>

    *   Search CPAN

        <http://search.cpan.org/dist/URI-Amazon-APA/>

ACKNOWLEDGEMENTS
    <http://docs.amazonwebservices.com/AWSECommerceService/latest/DG/index.h
    tml?rest-signature.html>

COPYRIGHT & LICENSE
    Copyright 2009 Dan Kogai, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
# Revision history for URI-Amazon-APA
#
# $Id: Changes,v 0.4 2011/05/21 21:53:23 dankogai Exp dankogai $
$Revision: 0.4 $ $Date: 2011/05/21 21:53:23 $
! lib/URI/Amazon/APA.pm
  Merged: #85910 (URI::query_form has potential to double encode)
  https://github.com/dankogai/p5-uri-amazon-apa/pull/1

0.04 2011/05/21 21:53:23
! lib/URI/Amazon/APA.pm
  Default API version upgraded to 2010-09-01
  https://rt.cpan.org/Ticket/Display.html?id=68244
! Makefile.PL
  Addressed: Undeclared dependency on Encode
  https://rt.cpan.org/Ticket/Display.html?id=50731

0.03 2009/10/19 09:33:51
! lib/URI/Amazon/APA.pm
  Single quote (') should be escaped
  http://github.com/miyagawa/p5-uri-amazon-apa/commit/857d4d3a99b2ad4d913f2da0205fbd81b07870b2
! lib/URI/Amazon/APA.pm
  No longer uses POSIX::strftime().
! Makefile.PL
  Requires URI.pm 1.36 or better
  http://coderepos.org/share/changeset/34935
  http://twitter.com/matsuu/status/3559718797

0.02 2009/05/22 19:31:05
! lib/URI/Amazon/APA.pm
  Handles non-ASCII query properly
  http://blog.livedoor.jp/dankogai/archives/51211577.html#comments

0.01    2009/05/10 11:01:30
        First version, released on an unsuspecting world.



lib/URI/Amazon/APA.pm
package URI::Amazon::APA;
use warnings;
use strict;
our $VERSION = sprintf "%d.%02d", q$Revision: 0.4 $ =~ /(\d+)/g;
use Carp;
use Digest::SHA qw(hmac_sha256_base64);
use URI::Escape;
use Encode qw/decode_utf8/;
use base 'URI::http';

sub new{
    my $class = shift;
    my $self  = URI->new(@_);
    ref $self eq 'URI::http' or carp "must be http";
    bless $self, $class;
}

sub sign {
    my $self  = shift;
    my (%arg) = @_;
    my %eq    = map { split /=/, $_ } split /&/, $self->query();
    my %q     = map { $_ => decode_utf8( uri_unescape( $eq{$_} ) ) } keys %eq;
    $q{Keywords} =~ s/\+/ /g if $q{Keywords};
    $q{AWSAccessKeyId} = $arg{key};
    $q{Timestamp} ||= do {
        my ( $ss, $mm, $hh, $dd, $mo, $yy ) = gmtime();
        join '',
          sprintf( '%04d-%02d-%02d', $yy + 1900, $mo + 1, $dd ), 'T',
          sprintf( '%02d:%02d:%02d', $hh,        $mm,     $ss ), 'Z';
    };
    $q{Version} ||= '2010-09-01';
    my $sq = join '&',
      map { $_ . '=' . uri_escape_utf8( $q{$_}, "^A-Za-z0-9\-_.~" ) }
      sort keys %q;
    my $tosign = join "\n", 'GET', $self->host, $self->path, $sq;
    my $signature = hmac_sha256_base64( $tosign, $arg{secret} );
    $signature .= '=' while length($signature) % 4;    # padding required
    $q{Signature} = $signature;
    $self->query_form( \%q );
    $self;
}

sub signature {
    my $self  = shift;
    my (%arg) = @_;
    my %eq = map { split /=/, $_ } split /&/, $self->query();
    my %q = map { $_ => uri_unescape( $eq{$_} ) } keys %eq;
    $q{Signature};
}

1; # End of URI::Amazon::APA

=head1 NAME

URI::Amazon::APA - URI to access Amazon Product Advertising API

=head1 VERSION

$Id: APA.pm,v 0.4 2011/05/21 21:53:23 dankogai Exp dankogai $

=head1 SYNOPSIS

  # self-explanatory
  use strict;
  use warnings;
  use URI::Amazon::APA;
  use LWP::UserAgent;
  use XML::Simple;
  use YAML::Syck;

  use URI::Amazon::APA; # instead of URI
  my $u = URI::Amazon::APA->new('http://webservices.amazon.com/onca/xml');
  $u->query_form(
    Service     => 'AWSECommerceService',
    Operation   => 'ItemSearch',
    Title       => shift || 'Perl',
    SearchIndex => 'Books',
  );
  $u->sign(
    key    => $public_key,
    secret => $private_key,
  );

  my $ua = LWP::UserAgent->new;
  my $r  = $ua->get($u);
  if ( $r->is_success ) {
    print YAML::Syck::Dump( XMLin( $r->content ) );
  }
  else {
    print $r->status_line, $r->as_string;
  }

=head1 EXPORT

None.

=head1 METHODS

This adds the following methods to L<URI> object

=head2 sign

Sings the URI accordingly to the Amazon Product Advertising API.

  $u->sign(
    key    => $public_key,
    secret => $private_key,
  );

=head2 signature

Checks the signature within the URI;

  print "The signature is " : $u->signature;

=head1 AUTHOR

Dan Kogai, C<< <dankogai at dan.co.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-uri-amazon-apa at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=URI-Amazon-APA>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc URI::Amazon::APA


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=URI-Amazon-APA>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/URI-Amazon-APA>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/URI-Amazon-APA>

=item * Search CPAN

L<http://search.cpan.org/dist/URI-Amazon-APA/>

=back

=head1 ACKNOWLEDGEMENTS

L<http://docs.amazonwebservices.com/AWSECommerceService/latest/DG/index.html?rest-signature.html>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Dan Kogai, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
